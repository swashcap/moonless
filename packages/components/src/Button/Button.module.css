/* Button */

.Button {
  border-style: solid;
  cursor: pointer;
  font-family: inherit;
  margin: 0;
  text-align: center;
  text-decoration: none;
}

.Button[disabled] {
  cursor: default;
}

/* Button: primary */

.ButtonPrimary {
  background: var(--component-button-primary-background-color-base);
  border-color: var(--component-button-primary-border-color-base);
  color: var(--component-button-primary-text-color-base);
}

.ButtonPrimary:hover,
:global(.storybook-hover).ButtonPrimary {
  background: var(--component-button-primary-background-color-hover);
  border-color: var(--component-button-primary-border-color-hover);
  color: var(--component-button-primary-text-color-hover);
}

.ButtonPrimary:focus,
:global(.storybook-focus).ButtonPrimary,
.ButtonPrimary:active,
:global(.storybook-active).ButtonPrimary {
  background: var(--component-button-primary-background-color-active);
  border-color: var(--component-button-primary-border-color-active);
  color: var(--component-button-primary-text-color-active);
}

.ButtonPrimary[disabled] {
  background: var(--component-button-primary-background-color-disabled);
  border-color: var(--component-button-primary-border-color-disabled);
  color: var(--component-button-primary-text-color-disabled);
}

/* Button: secondary */

.ButtonSecondary {
  background: var(--component-button-secondary-background-color-base);
  border-color: var(--component-button-secondary-border-color-base);
  color: var(--component-button-secondary-text-color-base);
}

.ButtonSecondary:hover,
:global(.storybook-hover).ButtonSecondary {
  background: var(--component-button-secondary-background-color-hover);
  border-color: var(--component-button-secondary-border-color-hover);
  color: var(--component-button-secondary-text-color-hover);
}

.ButtonSecondary:focus,
:global(.storybook-focus).ButtonSecondary,
.ButtonSecondary:active,
:global(.storybook-active).ButtonSecondary {
  background: var(--component-button-secondary-background-color-active);
  border-color: var(--component-button-secondary-border-color-active);
  color: var(--component-button-secondary-text-color-active);
}

.ButtonSecondary[disabled] {
  background: var(--component-button-secondary-background-color-disabled);
  border-color: var(--component-button-secondary-background-color-disabled);
  color: var(--component-button-secondary-text-color-disabled);
}

/** Button: Size */

.ButtonSmall {
  border-radius: var(--component-button-small-border-radius);
  border-width: var(--component-button-small-border-width);
  font-size: var(--component-button-small-font-size);
  line-height: var(--component-button-small-line-height);
  padding: var(--component-button-small-padding-vertical) var(--component-button-small-padding-horizontal);
}

.ButtonMedium {
  border-radius: var(--component-button-medium-border-radius);
  border-width: var(--component-button-medium-border-width);
  font-size: var(--component-button-medium-font-size);
  line-height: var(--component-button-medium-line-height);
  padding: var(--component-button-medium-padding-vertical) var(--component-button-medium-padding-horizontal);
}

.ButtonLarge {
  border-radius: var(--component-button-large-border-radius);
  border-width: var(--component-button-large-border-width);
  font-size: var(--component-button-large-font-size);
  line-height: var(--component-button-large-line-height);
  padding: var(--component-button-large-padding-vertical) var(--component-button-large-padding-horizontal);
}
