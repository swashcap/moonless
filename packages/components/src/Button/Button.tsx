import React from 'react';
import { clsx } from 'moonless-utils';

import styles from './Button.module.css';

export type ButtonSize = 'small' | 'medium' | 'large';

export type ButtonElement = HTMLAnchorElement | HTMLButtonElement;

export type ButtonVariant = 'primary' | 'secondary';

export interface ButtonOwnProps {
  size?: ButtonSize;
}

export interface ButtonAnchorProps
  extends React.AnchorHTMLAttributes<HTMLAnchorElement>,
    ButtonOwnProps {
  as: 'a';
}

export interface ButtonButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement>,
    ButtonOwnProps {
  as?: 'button';
}

export type ButtonProps<T extends ButtonElement> = T extends HTMLAnchorElement
  ? ButtonAnchorProps
  : ButtonButtonProps;

export type ButtonRef<T> = (
  instance: T | null
) => void | React.MutableRefObject<T | null> | null;

type BaseButtonProps<T extends ButtonElement> = ButtonProps<T> & {
  variant: ButtonVariant;
};

const BaseButtonFn = <T extends ButtonElement = HTMLButtonElement>(
  {
    as,
    children,
    className,
    size = 'medium',
    variant,
    ...rest
  }: BaseButtonProps<T>,
  ref: ButtonRef<T>
) => {
  const cls = clsx(
    styles.Button,
    size === 'small' && styles.ButtonSmall,
    size === 'medium' && styles.ButtonMedium,
    size === 'large' && styles.ButtonLarge,
    variant === 'primary' && styles.ButtonPrimary,
    variant === 'secondary' && styles.ButtonSecondary,
    className
  );

  if (as === 'a') {
    return (
      <a
        className={cls}
        ref={ref as React.LegacyRef<HTMLAnchorElement>}
        {...(rest as Omit<
          ButtonAnchorProps,
          'as' | 'children' | 'className' | 'size' | 'variant'
        >)}
      >
        {children}
      </a>
    );
  }

  return (
    <button
      className={cls}
      type="button"
      ref={ref as React.LegacyRef<HTMLButtonElement>}
      {...(rest as Omit<
        ButtonButtonProps,
        'as' | 'children' | 'className' | 'size' | 'variant'
      >)}
    >
      {children}
    </button>
  );
};

const BaseButton = React.forwardRef(
  BaseButtonFn as React.ForwardRefRenderFunction<
    ButtonElement,
    BaseButtonProps<ButtonElement>
  >
);

// Primary

/* eslint-disable-next-line @typescript-eslint/no-explicit-any */
const PrimaryButtonFn = (props: any, ref: any) => (
  <BaseButton ref={ref} variant="primary" {...props} />
);

export const PrimaryButton = React.forwardRef(
  PrimaryButtonFn as React.ForwardRefRenderFunction<
    ButtonElement,
    ButtonProps<ButtonElement>
  >
);

// Secondary

/* eslint-disable-next-line @typescript-eslint/no-explicit-any */
const SecondaryButtonFn = (props: any, ref: any) => (
  <BaseButton ref={ref} variant="secondary" {...props} />
);

export const SecondaryButton = React.forwardRef(
  SecondaryButtonFn as React.ForwardRefRenderFunction<
    ButtonElement,
    ButtonProps<ButtonElement>
  >
);
