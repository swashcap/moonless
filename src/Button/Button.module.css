/* Button */

.Button {
  border-style: solid;
  border-width: 1px;
  cursor: pointer;
  font-family: inherit;
  font-size: 100%;
  line-height: inherit;
  margin: 0;
  padding: 1em;
}

.Button[disabled] {
  cursor: default;
}

/* Button: primary */

.ButtonPrimary {
  background: navy;
  border-color: transparent;
  color: white;
}

/* stylelint-disable no-duplicate-selectors */
.ButtonPrimary:focus,
:global(.storybook-focus).ButtonPrimary {
  background: blue;
}

.ButtonPrimary:hover,
:global(.storybook-hover).ButtonPrimary {
  background: royalblue;
}

.ButtonPrimary:hover,
:global(.storybook-hover).ButtonPrimary {
  background: lightblue;
}
/* stylelint-enable no-duplicate-selectors */

.ButtonPrimary[disabled] {
  background: gray;
}

/* Button: secondary */

.ButtonSecondary {
  background: white;
  border-color: navy;
  color: navy;
}
