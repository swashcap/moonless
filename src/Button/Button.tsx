import React from 'react';

import { clsx } from '../utils/clsx';
import styles from './Button.module.css';

interface ButtonProps
  extends React.DetailedHTMLProps<
    React.ButtonHTMLAttributes<HTMLButtonElement>,
    HTMLButtonElement
  > {
  variant: 'primary' | 'secondary';
}

const Button: React.FC<ButtonProps> = ({ className, variant, ...rest }) => (
  <button
    className={clsx(
      styles.Button,
      variant === 'primary' && styles.ButtonPrimary,
      variant === 'secondary' && styles.ButtonSecondary,
      className
    )}
    {...rest}
  />
);

export type PrimaryButtonProps = Omit<ButtonProps, 'variant'>;

export const PrimaryButton: React.FC<PrimaryButtonProps> = (props) => (
  <Button variant="primary" {...props} />
);

export type SecondaryButtonProps = Omit<ButtonProps, 'variant'>;

export const SecondaryButton: React.FC<SecondaryButtonProps> = (props) => (
  <Button variant="secondary" {...props} />
);
