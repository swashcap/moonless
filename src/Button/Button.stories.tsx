import React from 'react';

import { PrimaryButton, SecondaryButton } from './Button';

export default {
  title: 'Component/Button',
};

export const Primary = () => (
  <div>
    <PrimaryButton>Primary</PrimaryButton>
    <PrimaryButton className="storybook-focus">focus</PrimaryButton>
    <PrimaryButton className="storybook-hover">hover</PrimaryButton>
    <PrimaryButton className="storybook-active">active</PrimaryButton>
    <PrimaryButton disabled>disabled</PrimaryButton>
  </div>
);

export const Secondary = () => (
  <div>
    <SecondaryButton>Secondary</SecondaryButton>
    <SecondaryButton className="storybook-focus">focus</SecondaryButton>
    <SecondaryButton className="storybook-hover">hover</SecondaryButton>
    <SecondaryButton className="storybook-active">active</SecondaryButton>
    <SecondaryButton disabled>disabled</SecondaryButton>
  </div>
);

export const OnClick = () => (
  <PrimaryButton onClick={console.log}>OnClick</PrimaryButton>
);

export const AsAnchor = () => (
  <div>
    <SecondaryButton as="a" href="#example">
      Go to example section
    </SecondaryButton>
    <PrimaryButton as="a" href="https://google.com" rel="nofollow">
      The Googs
    </PrimaryButton>
  </div>
);

export const Size = () => (
  <div>
    <PrimaryButton size="small">Small</PrimaryButton>
    <PrimaryButton size="medium">Medium</PrimaryButton>
    <PrimaryButton size="large">Large</PrimaryButton>
  </div>
);
