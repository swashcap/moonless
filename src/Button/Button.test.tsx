import React from 'react';
import { render } from '@testing-library/react';

import { ButtonSize, PrimaryButton, SecondaryButton } from './Button';

describe('Button', () => {
  describe.each([
    ['PrimaryButton', PrimaryButton],
    ['SecondaryButton', SecondaryButton],
  ])('%s', (name, Component) => {
    test('It should render.', () => {
      const { container } = render(<Component>Button</Component>);

      expect(container).toMatchSnapshot();
    });

    test('It should spread properties to the root element', () => {
      const { getByRole } = render(
        <Component className="test1" data-testid="test2">
          Test 3
        </Component>
      );

      const button = getByRole('button');

      expect(button).toHaveClass('test1');
      expect(button).toHaveAttribute('data-testid', 'test2');
      expect(button.textContent).toBe('Test 3');
    });

    test('It should render as an anchor', () => {
      const { getByRole } = render(
        <Component as="a" href="http://example.com">Go to example</Component>
      );

      expect(getByRole('link')).toHaveAttribute('href', 'http://example.com');
    });

    test.each<ButtonSize>(['small', 'medium', 'large'])(
      'It should render size %s.',
      (size) => {
        const { getByRole } = render(<Component size={size}>Button</Component>);

        expect(getByRole('button')).toHaveClass(
          `Button${size.replace(/(\w)/, (x) => x.toUpperCase())}`
        );
      }
    );

    test('It should forward ref.', () => {
      const ref = React.createRef<HTMLButtonElement>();
      const { getByRole } = render(<Component ref={ref}>Button</Component>);

      expect(getByRole('button')).toBe(ref.current);
    });
  });
});
