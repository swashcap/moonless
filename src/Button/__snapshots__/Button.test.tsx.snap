// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button PrimaryButton It should render. 1`] = `
<div>
  <button
    class="Button ButtonMedium ButtonPrimary"
    type="button"
  >
    Button
  </button>
</div>
`;

exports[`Button SecondaryButton It should render. 1`] = `
<div>
  <button
    class="Button ButtonMedium ButtonSecondary"
    type="button"
  >
    Button
  </button>
</div>
`;
